common/fixation_migrations.go

```diff
@@ -0,0 +1,77 @@
+package common
+
+import (
+	"fmt"
+	"math"
+
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	"github.com/lavanet/lava/common/types"
+)
+
+func prefixForErrors(from uint64) string {
+	return fmt.Sprintf("FixationStore: migration from version %d", from)
+}
+
+func (fs *FixationStore) MigrateVersion(ctx sdk.Context) (err error) {
+	from := fs.getVersion(ctx)
+	to := types.FixationVersion()
+
+	for from < to {
+		switch from {
+		case 1:
+			err = fs.migrate1to2(ctx)
+		default:
+			err = fmt.Errorf("%s not available", prefixForErrors(from))
+		}
+		if err != nil {
+			return err
+		}
+		from += 1
+	}
+
+	fs.setVersion(ctx, to)
+	return nil
+}
+
+// migrate1to2: fix refcounts
+//   - correct refcount of head (most recent) by adding one (because new
+//     entries used to begin with refcount 0 instead of refcount 1)
+//   - correct negative refcounts if found any (for extra care)
+func (fs *FixationStore) migrate1to2(ctx sdk.Context) error {
```

Comment:
why dont just use the GetEntry on all of the indexes instead of doing it manually.
also refcounts cant be zero since we have protection for it

Reply:
> why dont just use the GetEntry on all of the indexes instead of doing it manually.

For the head entries (i.e. most recent version of entries), indeed GetEntry() would suffice.
But we do three additional things:
- for entries with refcount zero and StaleAt not set, need to set the StaleAt (sanity check, shouldn't happen)
- for entries with StaleAt set: need to start a timer to ensure proper cleanup. (And if the StaleAt has already passed, set a timer for the next block - need to add that to the logic).
- for all entries, check that refcount is not too big - as it would mean underflow occurred before (sanity check, shouldn't happen).
- 
And for these last three items we must do the work manually.

> also refcounts cant be zero since we have protection for it

refcounts can be zero - for entries that are in the middle (i.e. not oldest). That's part of what the fixes of refcount fix.